src/VirtoCommerce.Platform.Web/Security/CustomUserManager.cs

```diff
@@ -174,21 +174,40 @@ public override async Task<IdentityResult> UpdateAsync(ApplicationUser user)
         {
             var result = await base.UpdateAsync(user);
 
-            if (result.Succeeded && user.Roles != null)
+            if (result.Succeeded)
             {
-                var targetRoles = (await GetRolesAsync(user));
-                var sourceRoles = user.Roles.Select(x => x.Name);
-
-                //Add
-                foreach (var newRole in sourceRoles.Except(targetRoles))
+                if (user.Roles != null)
                 {
-                    await AddToRoleAsync(user, newRole);
+                    var targetRoles = await GetRolesAsync(user);
+                    var sourceRoles = user.Roles.Select(x => x.Name);
+
+                    //Add
+                    foreach (var newRole in sourceRoles.Except(targetRoles))
+                    {
+                        await AddToRoleAsync(user, newRole);
+                    }
+
+                    //Remove
+                    foreach (var removeRole in targetRoles.Except(sourceRoles))
+                    {
+                        await RemoveFromRoleAsync(user, removeRole);
+                    }
                 }
 
-                //Remove
-                foreach (var removeRole in targetRoles.Except(sourceRoles))
+                if (user.Logins != null)
                 {
-                    await RemoveFromRoleAsync(user, removeRole);
+                    var targetLogins = await GetLoginsAsync(user);
+                    var sourceLogins = user.Logins.Select(x => new UserLoginInfo(x.LoginProvider, x.ProviderKey, null));
+
+                    foreach (var item in sourceLogins.Where(x => targetLogins.All(y => x.LoginProvider + x.ProviderKey != y.LoginProvider + y.ProviderKey)))
```

Comment:
Since the `UserLoginInfo` is derived from ValueObject you might use `.Equals()` for the more short-form of equality checking.

Reply:
UserLoginInfo is for AspNetCore.Identity, not a Valueobject